### PR TITLE
feat: track buffered outgoing messages

### LINKED ISSUE #12203: Track buffered outgoing p2p messages with a new metric
### Describe the feature

we have a buffer for outgoing messages:

https://github.com/paradigmxyz/reth/blob/ff9a42ae8fbf441f4fb9ca8dcea84c765b284721/crates/net/network/src/session/active.rs#L89-L90

but currently don't track how many outgoing messages we have in total.

we can easily add this by:
introducing a new helper type that wraps a new metric value https://github.com/paradigmxyz/reth/blob/ff9a42ae8fbf441f4fb9ca8dcea84c765b284721/crates/net/network/src/metrics.rs#L83-L83 this type should track additions and deletions from the sessions output buffer.
This must be instantiated as a clone here https://github.com/paradigmxyz/reth/blob/ff9a42ae8fbf441f4fb9ca8dcea84c765b284721/crates/net/network/src/session/mod.rs#L486-L486
so that every session shares the same value

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by caglaryucekaya**
Closes #12203

### CODE REVIEW

File: crates/net/network/src/session/active.rs [hunk 1]
```diff
@@ -757,6 +758,35 @@ fn calculate_new_timeout(current_timeout: Duration, estimated_rtt: Duration) ->
 
     smoothened_timeout.clamp(MINIMUM_TIMEOUT, MAXIMUM_TIMEOUT)
 }
+
+pub(crate) struct QueuedOutgoingMessages {
+    messages: VecDeque<OutgoingMessage>,
+    count: Gauge,
```
Comment (mattsse) [inline crates/net/network/src/session/active.rs:764]: looking at the ordering in Gaugefn impl for Atomic I believe this should work and not result in overflows

**Review by mattsse — CHANGES_REQUESTED** [review]
nits

File: crates/net/network/src/session/active.rs [hunk 2]
```diff
@@ -757,6 +758,35 @@ fn calculate_new_timeout(current_timeout: Duration, estimated_rtt: Duration) ->
...
+    pub const fn new(metric: Gauge) -> Self {
+        Self { messages: VecDeque::new(), count: metric }
+    }
+
+    pub fn push_back(&mut self, message: OutgoingMessage) {
+        self.messages.push_back(message);
+        self.count.increment(1);
+    }
+
+    pub fn pop_front(&mut self) -> Option<OutgoingMessage> {
```
[the remark this answers is not in the document]
Comment (caglaryucekaya) [reply crates/net/network/src/session/active.rs]: Done

File: crates/net/network/src/session/active.rs [hunk 3]
```diff
@@ -757,6 +758,35 @@ fn calculate_new_timeout(current_timeout: Duration, estimated_rtt: Duration) ->
...
+        let message = self.messages.pop_front();
+        if message.is_some() {
+            self.count.decrement(1);
+        }
```
[the remark this answers is not in the document]
Comment (caglaryucekaya) [reply crates/net/network/src/session/active.rs]: Done

File: crates/net/network/src/session/active.rs [hunk 4]
```diff
@@ -757,6 +758,35 @@ fn calculate_new_timeout(current_timeout: Duration, estimated_rtt: Duration) ->
 
     smoothened_timeout.clamp(MINIMUM_TIMEOUT, MAXIMUM_TIMEOUT)
 }
+
+pub(crate) struct QueuedOutgoingMessages {
+    messages: VecDeque<OutgoingMessage>,
+    count: Gauge,
+}
```
[the remark this answers is not in the document]
Comment (caglaryucekaya) [reply crates/net/network/src/session/active.rs:766]: Done

**Review by mattsse — APPROVED** [review]
lgtm

**Pull request merged**